# ¡¡¡ ANTES DE COMENZAR CAMBIA EL NOMBRE DE ESTE ARCHIVO !!!

Cambia el nombre del archivo de modo que empiece con las tres siguientes letras `P3-` y vaya seguido de los nombres de los alumnos separados de con un guión `-`. Para los espacios en blanco usar guiones bajos `_`

Ejemplo para un alumno:

`P3-Don_Quijote.ipynb`

Ejemplo para dos alumnos:

`P3-Don_Quijote-Sancho_Panza.ipynb`

# Recuerda que la entrega debe hacerse en formato `.ipynb`

# AWID 2

[AWID2 dataset source](https://icsdweb.aegean.gr/awid/awid2)

![Topology](https://icsdweb.aegean.gr/awid/static/media/blueprints.505abb3d.png)

In [1]:
import numpy as np
import pandas as pd

### Load Data

In [2]:
features = ['frame.interface_id','frame.dlt','frame.offset_shift','frame.time_epoch','frame.time_delta','frame.time_delta_displayed','frame.time_relative','frame.len','frame.cap_len','frame.marked','frame.ignored','radiotap.version','radiotap.pad','radiotap.length','radiotap.present.tsft','radiotap.present.flags','radiotap.present.rate','radiotap.present.channel','radiotap.present.fhss','radiotap.present.dbm_antsignal','radiotap.present.dbm_antnoise','radiotap.present.lock_quality','radiotap.present.tx_attenuation','radiotap.present.db_tx_attenuation','radiotap.present.dbm_tx_power','radiotap.present.antenna','radiotap.present.db_antsignal','radiotap.present.db_antnoise','radiotap.present.rxflags','radiotap.present.xchannel','radiotap.present.mcs','radiotap.present.ampdu','radiotap.present.vht','radiotap.present.reserved','radiotap.present.rtap_ns','radiotap.present.vendor_ns','radiotap.present.ext','radiotap.mactime','radiotap.flags.cfp','radiotap.flags.preamble','radiotap.flags.wep','radiotap.flags.frag','radiotap.flags.fcs','radiotap.flags.datapad','radiotap.flags.badfcs','radiotap.flags.shortgi','radiotap.datarate','radiotap.channel.freq','radiotap.channel.type.turbo','radiotap.channel.type.cck','radiotap.channel.type.ofdm','radiotap.channel.type.2ghz','radiotap.channel.type.5ghz','radiotap.channel.type.passive','radiotap.channel.type.dynamic','radiotap.channel.type.gfsk','radiotap.channel.type.gsm','radiotap.channel.type.sturbo','radiotap.channel.type.half','radiotap.channel.type.quarter','radiotap.dbm_antsignal','radiotap.antenna','radiotap.rxflags.badplcp','wlan.fc.type_subtype','wlan.fc.version','wlan.fc.type','wlan.fc.subtype','wlan.fc.ds','wlan.fc.frag','wlan.fc.retry','wlan.fc.pwrmgt','wlan.fc.moredata','wlan.fc.protected','wlan.fc.order','wlan.duration','wlan.ra','wlan.da','wlan.ta','wlan.sa','wlan.bssid','wlan.frag','wlan.seq','wlan.bar.type','wlan.ba.control.ackpolicy','wlan.ba.control.multitid','wlan.ba.control.cbitmap','wlan.bar.compressed.tidinfo','wlan.ba.bm','wlan.fcs_good','wlan_mgt.fixed.capabilities.ess','wlan_mgt.fixed.capabilities.ibss','wlan_mgt.fixed.capabilities.cfpoll.ap','wlan_mgt.fixed.capabilities.privacy','wlan_mgt.fixed.capabilities.preamble','wlan_mgt.fixed.capabilities.pbcc','wlan_mgt.fixed.capabilities.agility','wlan_mgt.fixed.capabilities.spec_man','wlan_mgt.fixed.capabilities.short_slot_time','wlan_mgt.fixed.capabilities.apsd','wlan_mgt.fixed.capabilities.radio_measurement','wlan_mgt.fixed.capabilities.dsss_ofdm','wlan_mgt.fixed.capabilities.del_blk_ack','wlan_mgt.fixed.capabilities.imm_blk_ack','wlan_mgt.fixed.listen_ival','wlan_mgt.fixed.current_ap','wlan_mgt.fixed.status_code','wlan_mgt.fixed.timestamp','wlan_mgt.fixed.beacon','wlan_mgt.fixed.aid','wlan_mgt.fixed.reason_code','wlan_mgt.fixed.auth.alg','wlan_mgt.fixed.auth_seq','wlan_mgt.fixed.category_code','wlan_mgt.fixed.htact','wlan_mgt.fixed.chanwidth','wlan_mgt.fixed.fragment','wlan_mgt.fixed.sequence','wlan_mgt.tagged.all','wlan_mgt.ssid','wlan_mgt.ds.current_channel','wlan_mgt.tim.dtim_count','wlan_mgt.tim.dtim_period','wlan_mgt.tim.bmapctl.multicast','wlan_mgt.tim.bmapctl.offset','wlan_mgt.country_info.environment','wlan_mgt.rsn.version','wlan_mgt.rsn.gcs.type','wlan_mgt.rsn.pcs.count','wlan_mgt.rsn.akms.count','wlan_mgt.rsn.akms.type','wlan_mgt.rsn.capabilities.preauth','wlan_mgt.rsn.capabilities.no_pairwise','wlan_mgt.rsn.capabilities.ptksa_replay_counter','wlan_mgt.rsn.capabilities.gtksa_replay_counter','wlan_mgt.rsn.capabilities.mfpr','wlan_mgt.rsn.capabilities.mfpc','wlan_mgt.rsn.capabilities.peerkey','wlan_mgt.tcprep.trsmt_pow','wlan_mgt.tcprep.link_mrg','wlan.wep.iv','wlan.wep.key','wlan.wep.icv','wlan.tkip.extiv','wlan.ccmp.extiv','wlan.qos.tid','wlan.qos.priority','wlan.qos.eosp','wlan.qos.ack','wlan.qos.amsdupresent','wlan.qos.buf_state_indicated','wlan.qos.bit4','wlan.qos.txop_dur_req','_','data.len','class']
dtypes = {'frame.interface_id': np.int64, 'frame.dlt': np.float64, 'frame.offset_shift': np.float64, 'frame.time_epoch': np.float64, 'frame.time_delta': np.float64, 'frame.time_delta_displayed': np.float64, 'frame.time_relative': np.float64, 'frame.len': np.int64, 'frame.cap_len': np.int64, 'frame.marked': np.int64, 'frame.ignored': np.int64, 'radiotap.version': np.int64, 'radiotap.pad': np.int64, 'radiotap.length': np.int64, 'radiotap.present.tsft': np.int64, 'radiotap.present.flags': np.int64, 'radiotap.present.rate': np.int64, 'radiotap.present.channel': np.int64, 'radiotap.present.fhss': np.int64, 'radiotap.present.dbm_antsignal': np.int64, 'radiotap.present.dbm_antnoise': np.int64, 'radiotap.present.lock_quality': np.int64, 'radiotap.present.tx_attenuation': np.int64, 'radiotap.present.db_tx_attenuation': np.int64, 'radiotap.present.dbm_tx_power': np.int64, 'radiotap.present.antenna': np.int64, 'radiotap.present.db_antsignal': np.int64, 'radiotap.present.db_antnoise': np.int64, 'radiotap.present.rxflags': np.int64, 'radiotap.present.xchannel': np.int64, 'radiotap.present.mcs': np.int64, 'radiotap.present.ampdu': np.int64, 'radiotap.present.vht': np.int64, 'radiotap.present.reserved': str, 'radiotap.present.rtap_ns': np.int64, 'radiotap.present.vendor_ns': np.int64, 'radiotap.present.ext': np.int64, 'radiotap.mactime': np.float64, 'radiotap.flags.cfp': np.float64, 'radiotap.flags.preamble': np.float64, 'radiotap.flags.wep': np.float64, 'radiotap.flags.frag': np.float64, 'radiotap.flags.fcs': np.float64, 'radiotap.flags.datapad': np.float64, 'radiotap.flags.badfcs': np.float64, 'radiotap.flags.shortgi': np.float64, 'radiotap.datarate': np.float64, 'radiotap.channel.freq': np.float64, 'radiotap.channel.type.turbo': np.float64, 'radiotap.channel.type.cck': np.float64, 'radiotap.channel.type.ofdm': np.float64, 'radiotap.channel.type.2ghz': np.float64, 'radiotap.channel.type.5ghz': np.float64, 'radiotap.channel.type.passive': np.float64, 'radiotap.channel.type.dynamic': np.float64, 'radiotap.channel.type.gfsk': np.float64, 'radiotap.channel.type.gsm': np.float64, 'radiotap.channel.type.sturbo': np.float64, 'radiotap.channel.type.half': np.float64, 'radiotap.channel.type.quarter': np.float64, 'radiotap.dbm_antsignal': np.float64, 'radiotap.antenna': np.float64, 'radiotap.rxflags.badplcp': np.float64, 'wlan.fc.type_subtype': str, 'wlan.fc.version': np.int64, 'wlan.fc.type': np.int64, 'wlan.fc.subtype': np.int64, 'wlan.fc.ds': str, 'wlan.fc.frag': np.int64, 'wlan.fc.retry': np.int64, 'wlan.fc.pwrmgt': np.int64, 'wlan.fc.moredata': np.int64, 'wlan.fc.protected': np.int64, 'wlan.fc.order': np.int64, 'wlan.duration': np.float64, 'wlan.ra': str, 'wlan.da': str, 'wlan.ta': str, 'wlan.sa': str, 'wlan.bssid': str, 'wlan.frag': np.float64, 'wlan.seq': np.float64, 'wlan.bar.type': str, 'wlan.ba.control.ackpolicy': np.float64, 'wlan.ba.control.multitid': np.float64, 'wlan.ba.control.cbitmap': np.float64, 'wlan.bar.compressed.tidinfo': str, 'wlan.ba.bm': str, 'wlan.fcs_good': np.float64, 'wlan_mgt.fixed.capabilities.ess': np.float64, 'wlan_mgt.fixed.capabilities.ibss': np.float64, 'wlan_mgt.fixed.capabilities.cfpoll.ap': str, 'wlan_mgt.fixed.capabilities.privacy': np.float64, 'wlan_mgt.fixed.capabilities.preamble': np.float64, 'wlan_mgt.fixed.capabilities.pbcc': np.float64, 'wlan_mgt.fixed.capabilities.agility': np.float64, 'wlan_mgt.fixed.capabilities.spec_man': np.float64, 'wlan_mgt.fixed.capabilities.short_slot_time': np.float64, 'wlan_mgt.fixed.capabilities.apsd': np.float64, 'wlan_mgt.fixed.capabilities.radio_measurement': np.float64, 'wlan_mgt.fixed.capabilities.dsss_ofdm': np.float64, 'wlan_mgt.fixed.capabilities.del_blk_ack': np.float64, 'wlan_mgt.fixed.capabilities.imm_blk_ack': np.float64, 'wlan_mgt.fixed.listen_ival': str, 'wlan_mgt.fixed.current_ap': str, 'wlan_mgt.fixed.status_code': str, 'wlan_mgt.fixed.timestamp': str, 'wlan_mgt.fixed.beacon': np.float64, 'wlan_mgt.fixed.aid': str, 'wlan_mgt.fixed.reason_code': str, 'wlan_mgt.fixed.auth.alg': np.float64, 'wlan_mgt.fixed.auth_seq': str, 'wlan_mgt.fixed.category_code': np.float64, 'wlan_mgt.fixed.htact': str, 'wlan_mgt.fixed.chanwidth': str, 'wlan_mgt.fixed.fragment': np.float64, 'wlan_mgt.fixed.sequence': np.float64, 'wlan_mgt.tagged.all': np.float64, 'wlan_mgt.ssid': str, 'wlan_mgt.ds.current_channel': np.float64, 'wlan_mgt.tim.dtim_count': np.float64, 'wlan_mgt.tim.dtim_period': np.float64, 'wlan_mgt.tim.bmapctl.multicast': np.float64, 'wlan_mgt.tim.bmapctl.offset': str, 'wlan_mgt.country_info.environment': str, 'wlan_mgt.rsn.version': np.float64, 'wlan_mgt.rsn.gcs.type': np.float64, 'wlan_mgt.rsn.pcs.count': np.float64, 'wlan_mgt.rsn.akms.count': np.float64, 'wlan_mgt.rsn.akms.type': np.float64, 'wlan_mgt.rsn.capabilities.preauth': np.float64, 'wlan_mgt.rsn.capabilities.no_pairwise': np.float64, 'wlan_mgt.rsn.capabilities.ptksa_replay_counter': str, 'wlan_mgt.rsn.capabilities.gtksa_replay_counter': str, 'wlan_mgt.rsn.capabilities.mfpr': np.float64, 'wlan_mgt.rsn.capabilities.mfpc': np.float64, 'wlan_mgt.rsn.capabilities.peerkey': np.float64, 'wlan_mgt.tcprep.trsmt_pow': np.float64, 'wlan_mgt.tcprep.link_mrg': np.float64, 'wlan.wep.iv': str, 'wlan.wep.key': np.float64, 'wlan.wep.icv': str, 'wlan.tkip.extiv': str, 'wlan.ccmp.extiv': str, 'wlan.qos.tid': np.float64, 'wlan.qos.priority': np.float64, 'wlan.qos.eosp': np.float64, 'wlan.qos.ack': str, 'wlan.qos.amsdupresent': np.float64, 'wlan.qos.buf_state_indicated': np.float64, 'wlan.qos.bit4': np.float64, 'wlan.qos.txop_dur_req': np.float64, '_': np.float64, 'data.len': np.float64, 'class': str}

awid_train = pd.read_csv("./dataset/AWID-CLS-R-Trn.csv", header=None, names=features, dtype=dtypes, na_values='?')
awid_test = pd.read_csv("./dataset/AWID-CLS-R-Tst.csv", header=None, names=features, dtype=dtypes, na_values='?')

### Convert problem to binary: normal vs attack (OPTIONAL)

In [3]:
awid_train['class'] = awid_train['class'].apply(lambda x: 'normal' if x == 'normal' else 'attack')
awid_test['class'] = awid_test['class'].apply(lambda x: 'normal' if x == 'normal' else 'attack')

### Transform categorical features to ordinal values so most ml algorithms can be applied


In [4]:
from sklearn.preprocessing import OrdinalEncoder

# Columns with str dtype are categorical.
categorical_columns = [column for column, column_type in dtypes.items() if column_type is str]

# Missing values are encoded as -1. Not known values during test transformation are also encoded as -1
encoder = OrdinalEncoder(encoded_missing_value=-1, handle_unknown='use_encoded_value', unknown_value=-1)

# Apply Ordinal Encoder
awid_train[categorical_columns] = encoder.fit_transform(awid_train[categorical_columns])
awid_test[categorical_columns] = encoder.transform(awid_test[categorical_columns])

In [5]:
X_awid_train = awid_train.iloc[:,:-1]
y_awid_train = awid_train['class']
X_awid_test = awid_test.iloc[:,:-1]
y_awid_test = awid_test['class']

### Proyecto de aprendizaje automático aplicado a ciberseguridad

Entrena un modelo para que sea capaz de detectar los ataques de un intruso en una red wifi de una oficina cuyo tráfico es monitorizado.

Para la evaluación de este modelo se tendrá en cuenta:
- Los pasos seguidos para procesar los datos y entrenar el modelo (7 puntos de 10). Los pasos deberán estár debidamente explicados.
- El rendimiento del modelo en la detección de ataques. (3 puntos de 10). Se utilizará el `balanced_accuracy_score` como métrica para evaluar el rendimiento del modelo en la detección de ataques frente a tráfico normal. Esta métrica se debe aplicar sobre etiquetas binarias: ataque vs tráfico normal.

## Preprocesamiento de Datos

Primero eliminamos las columnas cuya varianza sea 0

In [6]:
#Eliminacion de columnas cuya varianza sea 0
varianzas = np.var(X_awid_train, axis= 0)
columnas_eliminadas = varianzas[varianzas==0].index
X_awid_train = X_awid_train.drop(columns= columnas_eliminadas)
X_awid_test = X_awid_test.drop(columns= columnas_eliminadas)

In [7]:
columnas_eliminadas

Index(['frame.interface_id', 'frame.offset_shift', 'frame.marked',
       'frame.ignored', 'radiotap.version', 'radiotap.pad',
       'radiotap.present.rate', 'radiotap.present.fhss',
       'radiotap.present.dbm_antnoise', 'radiotap.present.lock_quality',
       'radiotap.present.tx_attenuation', 'radiotap.present.db_tx_attenuation',
       'radiotap.present.dbm_tx_power', 'radiotap.present.db_antsignal',
       'radiotap.present.db_antnoise', 'radiotap.present.xchannel',
       'radiotap.present.mcs', 'radiotap.present.ampdu',
       'radiotap.present.vht', 'radiotap.present.reserved',
       'radiotap.present.rtap_ns', 'radiotap.present.vendor_ns',
       'radiotap.present.ext', 'radiotap.flags.cfp', 'radiotap.flags.preamble',
       'radiotap.flags.wep', 'radiotap.flags.frag', 'radiotap.flags.fcs',
       'radiotap.flags.datapad', 'radiotap.flags.badfcs',
       'radiotap.flags.shortgi', 'radiotap.channel.type.turbo',
       'radiotap.channel.type.2ghz', 'radiotap.channel.type.5ghz

Despues eliminamos las columnas que sean todo nulo

In [8]:
#Eliminacion de columnas cuya columna entera sea nula
columnas_null = X_awid_train.columns[X_awid_train.isnull().all()]
X_awid_train = X_awid_train.drop(columns= columnas_null)
X_awid_test = X_awid_test.drop(columns= columnas_null)

In [9]:
columnas_null

Index(['frame.dlt', 'wlan.qos.buf_state_indicated'], dtype='object')

Se rellenan valores nulos. -1 en columnas categoricas y la media en las columnas numericas

In [10]:
#Relleno los valores nulos con -1
categorical_columns = [col for col in categorical_columns if col in X_awid_train]
numerical_columns = X_awid_train.select_dtypes(include=[np.number]).columns
mean = X_awid_train[numerical_columns].mean()

X_awid_train[categorical_columns] = X_awid_train[categorical_columns].fillna(-1)
X_awid_test[categorical_columns] = X_awid_test[categorical_columns].fillna(-1)

X_awid_train[numerical_columns] = X_awid_train[numerical_columns].fillna(mean)
X_awid_test[numerical_columns] = X_awid_test[numerical_columns].fillna(mean)

In [11]:
print(X_awid_train.shape)
print(X_awid_test.shape)

(1795575, 96)
(575643, 96)


## Normalizacion de los datos

In [12]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

X_awid_train_norm = pd.DataFrame(scaler.fit_transform(X_awid_train),columns=X_awid_train.columns, index= X_awid_train.index)
X_awid_test_norm = pd.DataFrame(scaler.transform(X_awid_test),columns=X_awid_test.columns, index= X_awid_test.index)


## Buscamos usando distintos modelos cual nos dara el mejor resultado

Sabemos que tenemos que aplicar metodologia de clasificación ya que tenemos que clasificar los datos entre trafico normal y trafico malicioso

In [13]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import balanced_accuracy_score

## Prueba 1 - KNN

kNN es un metodo basado en la simitud entre datos. Para ellos tenemos que seleccion un valor de k que sera el numero de puntos cercanos que se consideran para hacer la prediccion. Para ello se mide la distancia entre los datos y se identifican los que tiene mas cercanos.

Despues, se asigna una clase entre los k vecinos

In [19]:
#TODO
# Apliquemos Knn
from sklearn.neighbors import KNeighborsClassifier

# Definir el modelo de clasificacion K-nn
neigh = KNeighborsClassifier(3)

# Entrenarlo
neigh.fit(X_awid_train_norm,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = neigh.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

MSE : 0.10616996992927909
Accuracy: 0.8238549004845538


Precision de 82.38%
kNN es muy lento y no es lo mas eficaz para volumen altos de datos. Por ellos pasaremos a utilizar metodos de ensemble.

Los métodos de ensemble combinan múltiples modelos de aprendizaje automático para obtener predicciones más precisas y robustas. La idea principal es que varios modelos juntos pueden reducir errores, mejorar la generalización y minimizar el sobreajuste. Hay 3 tipos de ensemble pero nosotros solo vamos a aplicar 2 de ellos que son:

**Bagging**:
- Entrena múltiples modelos independientes en subconjuntos aleatorios de los datos.
- Los modelos son del mismo tipo.
- La predicción final se obtiene por votación.
- Reduce la varianza, útil para modelos que sobreajustan.
- Ejemplos:
    - RandomForestClassifier
    - BaggingClassifier

**Boosting**:
- Los modelos se entrenan secuencialmente, cada uno intentando corregir los errores del anterior.
- Da mayor peso a las observaciones mal clasificadas.
- Reduce el sesgo y mejora la precisión, pero puede ser más propenso a sobreajuste.
- Ejemplos:
    - GradientBoostingClassifier
    - AdaBoostClassifier

## Prueba 2 - Ensembles de Bagging

### Prueba 3.1 - Bagging

In [14]:
# Apliquemos Ensemble de Bagging con BaggingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import mean_squared_error, balanced_accuracy_score

#### Datos no normalizados

El Bagging Classifier (Bootstrap Aggregating Classifier) es una técnica de aprendizaje de ensamblado (ensemble learning) que mejora la estabilidad y precisión de los modelos de machine learning al reducir la varianza. Se encuentra implementado en Scikit-learn dentro del módulo sklearn.ensemble

**Uso de sus hiperparámetros**
- estimator: especifíca el modelo base (decisionTreeClassifier)
- n_estimator: número de clasificadores base a entrenar.
- random_state: es clave en machine learning porque permite controlar la aleatoriedad en el entrenamiento de modelos. En el contexto de BaggingClassifier y otros algoritmos, define la semilla del generador de números aleatorios, asegurando que los resultados sean reproducibles.

In [15]:
# Definir el modelo base
base_model = DecisionTreeClassifier(max_depth=1)

# Definir el modelo de Bagging
model = BaggingClassifier(estimator=base_model, n_estimators=100, random_state=42)

print("Training: ")
# Entrenar el modelo
model.fit(X_awid_train, y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test)

# Obtener el error cuadrático medio y la precisión
MSE_tst = mean_squared_error(y_awid_test, y_est)
accuracy = balanced_accuracy_score(y_awid_test, y_est)

# Imprimir el error y precisión del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.04918326115318001
Accuracy: 0.8478377974809932


Se obtiene un 84.78% lo que no es un gran porcentaje de acierto. Esto se puede deber a que no elimina el sesgo del modelo base ya que es un algoritmo de bagging. Además, es computacionalmente costoso

#### Datos Normalizados

In [16]:
# Definir el modelo base
base_model = DecisionTreeClassifier(max_depth=1)

# Definir el modelo de Bagging
model = BaggingClassifier(estimator=base_model, n_estimators=100, random_state=42)

print("Training: ")
# Entrenar el modelo
model.fit(X_awid_train_norm, y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadrático medio y la precisión
MSE_tst = mean_squared_error(y_awid_test, y_est)
accuracy = balanced_accuracy_score(y_awid_test, y_est)

# Imprimir el error y precisión del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.04918326115318001
Accuracy: 0.8478377974809932


Al realizar el mismo modelo pero con los datos normalizados, vemos que el porcentaje de acierto se queda en un 84.78%. Luego la normalización de los datos no ayuda a la precisión del modelo.

**Los árboles de decisión no son sensibles a la escala de los datos porque:**

- Hacen divisiones binarias en función de valores de características individuales, sin importar sus magnitudes absolutas.
- No usan distancia euclidiana ni cálculos basados en varianzas.
- Cada variable es evaluada de forma independiente para encontrar el mejor punto de división.

### Prueba 3.2 - Random Forest

In [16]:
from sklearn.ensemble import RandomForestClassifier

In [17]:
# Definir el modelo de Random Forest
model = RandomForestClassifier(n_estimators=100, max_depth=1, random_state=42)

print("Training: ")
# Entrenar el modelo
model.fit(X_awid_train, y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test)

# Obtener el error cuadrático medio y la precisión
MSE_tst = mean_squared_error(y_awid_test, y_est)
accuracy = balanced_accuracy_score(y_awid_test, y_est)

# Imprimir el error y precisión del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.0779267705852412
Accuracy: 0.5


#### Datos Normalizados

In [18]:
# Definir el modelo de Random Forest
model = RandomForestClassifier(n_estimators=100, max_depth=1, random_state=42)

print("Training: ")
# Entrenar el modelo
model.fit(X_awid_train_norm, y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadrático medio y la precisión
MSE_tst = mean_squared_error(y_awid_test, y_est)
accuracy = balanced_accuracy_score(y_awid_test, y_est)

# Imprimir el error y precisión del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.0779267705852412
Accuracy: 0.5


## Prueba 3 - Ensembles de Boosting

### Prueba 3.1 - Ada Boosting

AdaBoost es un método de ensemble que combina múltiples clasificadores débiles ponderándolos de manera iterativa para mejorar la precisión global del modelo. 
Este funciona asignado los mismo pesos a todas las muestras de entrenamiento al principio. Despues de cada iteración, se le asigna mayor peso a aquellas muestras que fueron clasificadas incorrectamente. La precision final es la combinacion ponderada de todos los clasificadores.

Para ello hemos implementado un arbol de decisiones como el modelo base(clasificador débil). Este modelo funciona construyendo el arbol de decisiones donde cada nodo es una caracteristica es una caracteristica del conjuntos de datos.

Parametros de DecisionTreeClassifier que vamos a modificar:
- **max_depth**: Profundidad maxima que se recorrera del arbol de decisiones

Parametros de AdaBoostClassifier que vamos a modificar:
- **estimator**: El clasificador base que vamos a usar
- **n_estimators**: El número de clasificadores debiles que vamos a usar 
- **learning_rate**: El peso aplicado a cada clasificador cada iteración

In [15]:
# Apliquemos Ensemble de Boosting con AdaBoosting
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier

#### Datos sin normalizar
Primero probaremos con los datos sin normalizar

In [ ]:
# Definir el modelo de clasificacion de AdaBoosting
base_model = DecisionTreeClassifier(max_depth=1)
model = AdaBoostClassifier(estimator= base_model, n_estimators=100, learning_rate=0.5, random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.024824413742545293
Accuracy: 0.8895368931448729


Precision de 88,95%. Es una buena precision para comenzar pero deberiamos intentar acercarnos al 100% por eso probaremos con los datos normalizados ahora

#### Datos Normalizados

In [20]:
# Definir el modelo de clasificacion de AdaBoosting
base_model = DecisionTreeClassifier(max_depth=1)
model = AdaBoostClassifier(estimator= base_model, n_estimators=100, learning_rate=0.5, random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train_norm,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.024824413742545293
Accuracy: 0.8895368931448729


Precision de 88,95%. La precision no ha cambiado asi que ahora tocar ir tocando ciertos hiperparametros para ver si conseguimos mejorar la precision

#### Cambiamos ciertos hiperparametros para ver si conseguimos subir la precision
Aumentamos la profundidad del modelo base y bajamos el learning rate

In [ ]:
# Definir el modelo de clasificacion de AdaBoosting
base_model = DecisionTreeClassifier(max_depth=3) 
model = AdaBoostClassifier(estimator= base_model, n_estimators=100, learning_rate=0.1, random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train_norm,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.017814860946802097
Accuracy: 0.9082973458033456


Precision de 90,83%. Es un resultado mejor pero puede que este habiendo sobre aprendizaje asi que bajaremos el numero de estimadores

In [18]:
# Definir el modelo de clasificacion de AdaBoosting
base_model = DecisionTreeClassifier(max_depth=3)
model = AdaBoostClassifier(estimator= base_model, n_estimators=50, learning_rate=0.1, random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.01399478496220748
Accuracy: 0.9358080830954062


Precision de 93,58%. Hemos conseguido una buena precision con este modelo pero vamos a probar otro ensemble de Boosting para ver si mejoramos la precision aun mas

### Prueba 3.2 - Gradient Boosting

GradientBoosting es otro metodo de ensemble que tambien esta basado en arboles de decision. Este en vez de ajustar pesos, intenta minimizar el error ajustando el modelo en funcion del gradiente de la función de perdida. Primero, se empieza con un modelo inicial simple y se calculan los residuos entre la prediccion y la realidad. Despues, se van construyendo nuevos modelos para predecir los residuos y nuevas predicciones sumando los nuevos modelos y ajustando una taza de aprendizaje

Ventajas:
- Mas robusto frente a valores atipicos
- Funciona bien frente a datos complejos y grandes volumenes de datos
- Mas flexible

Desventajas:
- Mayor probabilidad de sobreajuste si no se ajustan bien los parametros
- Mas lento que AdaBoost

Parametros a modificar:
- **learning_rate**: Empequeñece la contribucion de cada arbol de decision
- **n_estimators**: Numero de arboles
- **max_depth**: Profundidad maxima de cada arbol

In [16]:
# Apliquemos Ensemble de Boosting con GradientBoosting
from sklearn.ensemble import GradientBoostingClassifier

#### Datos Normalizados
Esta vez empezamos con los datos normalizados

In [ ]:
# Definir el modelo de clasificacion de GradientBoosting
model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1,max_depth=3,random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train_norm,y_awid_train)

print("Trained: ")
# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
Trained: 
MSE : 0.015487029287249215
Accuracy: 0.9031207246753497


Precision del 90,31%. Veo si añadiendo mas profundida consigo mejorar mis resultados

#### Veo que cambia si añado mas profundidad

In [ ]:
# Definir el modelo de clasificacion de GradientBoosting
model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1,max_depth=5,random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train_norm,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.05368952632100104
Accuracy: 0.7253000422592593


Precision de 72,53%. Ha habido sobre aprendizaje. Vuelvo a probar con los mismos parametros anteriores pero con los datos sin normalizar

#### Pruebo entrenando con datos no normalizados

In [ ]:
# Definir el modelo de clasificacion de GradientBoosting
model = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1,max_depth=3,random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.012959421030048138
Accuracy: 0.9289510020727874


#### Busco los mejores hiperparametros

Al principio pensamos en usar GridSearch para buscar los mejores parametros. Sin embargo nos dimos cuenta que necesitamos una solucion mas rapida ya que GridSearch a pesar de ser muy precisa, era muy lenta. Por ello pasamos a usar Optuna.

Para poder usar optuna, primero tenemos que definir el espacio de busqueda estableciendo los hiperparametros a modificar y sus rangos. Despues se van sugeriendo hiperparametros y entrenando modelos para ver cual da mayor precision.

Funciona por un sistema de trails en el cuando se puede ejecutar simultaneamente haciendo la busqueda mas eficiente

In [17]:
import optuna
from sklearn.model_selection import cross_val_score

c:\U-TAD\Segundo Carrera\Proyectos II\Cyber\.venv\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [ ]:
def objective(trial):
    parametros = {
        "n_estimators": [50, 100],
        "learning_rate": [0.01, 0.1, 0.3, 0.5],
        "max_depth": [3, 5, 7]
    }
    
    n_estimators = trial.suggest_categorical("n_estimators", parametros["n_estimators"])
    learning_rate = trial.suggest_categorical("learning_rate", parametros["learning_rate"])
    max_depth = trial.suggest_categorical("max_depth", parametros["max_depth"])
    
    model = GradientBoostingClassifier(
        n_estimators=n_estimators,
        learning_rate=learning_rate,
        max_depth=max_depth,
        random_state=42
    )
    
    score = cross_val_score(model, X_awid_train_norm, y_awid_train, cv=5, scoring='balanced_accuracy').mean()
    print(f"Trial {trial.number}: Accuracy = {score}")
    return score

# Run Optuna optimization
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=20,n_jobs=-1)

# Best hyperparameters
best_params = study.best_params
print("Best Parameters:", best_params)

[I 2025-03-17 00:52:46,844] A new study created in memory with name: no-name-bc3dcf7f-3e94-4bd3-8fb3-ac7530a01864
[I 2025-03-17 02:10:22,412] Trial 10 finished with value: 0.9574059408151209 and parameters: {'n_estimators': 50, 'learning_rate': 0.1, 'max_depth': 3}. Best is trial 10 with value: 0.9574059408151209.


Trial 10: Accuracy = 0.9574059408151209


[I 2025-03-17 02:15:14,636] Trial 7 finished with value: 0.5 and parameters: {'n_estimators': 50, 'learning_rate': 0.01, 'max_depth': 3}. Best is trial 10 with value: 0.9574059408151209.


Trial 7: Accuracy = 0.5


[I 2025-03-17 02:16:21,539] Trial 15 finished with value: 0.9845842587594351 and parameters: {'n_estimators': 50, 'learning_rate': 0.5, 'max_depth': 3}. Best is trial 15 with value: 0.9845842587594351.


Trial 15: Accuracy = 0.9845842587594351


[I 2025-03-17 02:57:39,539] Trial 11 finished with value: 0.8419472487779004 and parameters: {'n_estimators': 50, 'learning_rate': 0.3, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.


Trial 11: Accuracy = 0.8419472487779004


[I 2025-03-17 03:02:53,144] Trial 0 finished with value: 0.5 and parameters: {'n_estimators': 50, 'learning_rate': 0.01, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.


Trial 0: Accuracy = 0.5


[I 2025-03-17 03:26:18,432] Trial 13 finished with value: 0.7690714400756989 and parameters: {'n_estimators': 100, 'learning_rate': 0.01, 'max_depth': 3}. Best is trial 15 with value: 0.9845842587594351.


Trial 13: Accuracy = 0.7690714400756989


[I 2025-03-17 03:31:34,734] Trial 8 finished with value: 0.9417173226107215 and parameters: {'n_estimators': 100, 'learning_rate': 0.5, 'max_depth': 3}. Best is trial 15 with value: 0.9845842587594351.


Trial 8: Accuracy = 0.9417173226107215


[I 2025-03-17 03:37:58,140] Trial 5 finished with value: 0.9043621013095272 and parameters: {'n_estimators': 50, 'learning_rate': 0.1, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 5: Accuracy = 0.9043621013095272


[I 2025-03-17 03:41:27,383] Trial 1 finished with value: 0.9043621013095272 and parameters: {'n_estimators': 50, 'learning_rate': 0.1, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 1: Accuracy = 0.9043621013095272


[I 2025-03-17 03:42:17,123] Trial 6 finished with value: 0.9043621013095272 and parameters: {'n_estimators': 50, 'learning_rate': 0.1, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 6: Accuracy = 0.9043621013095272


[I 2025-03-17 03:45:14,967] Trial 14 finished with value: 0.9142905262579619 and parameters: {'n_estimators': 50, 'learning_rate': 0.5, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 14: Accuracy = 0.9142905262579619


[I 2025-03-17 04:05:08,565] Trial 17 finished with value: 0.5 and parameters: {'n_estimators': 50, 'learning_rate': 0.01, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.


Trial 17: Accuracy = 0.5


[I 2025-03-17 04:27:45,025] Trial 16 finished with value: 0.9509712892776339 and parameters: {'n_estimators': 50, 'learning_rate': 0.3, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 16: Accuracy = 0.9509712892776339


[I 2025-03-17 04:28:48,275] Trial 3 finished with value: 0.9123792080675137 and parameters: {'n_estimators': 100, 'learning_rate': 0.3, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.


Trial 3: Accuracy = 0.9123792080675137


[I 2025-03-17 04:28:59,068] Trial 19 finished with value: 0.8419472487779004 and parameters: {'n_estimators': 50, 'learning_rate': 0.3, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.


Trial 19: Accuracy = 0.8419472487779004


[I 2025-03-17 04:29:29,518] Trial 12 finished with value: 0.9123792080675137 and parameters: {'n_estimators': 100, 'learning_rate': 0.3, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.
[I 2025-03-17 04:29:29,684] Trial 4 finished with value: 0.9123792080675137 and parameters: {'n_estimators': 100, 'learning_rate': 0.3, 'max_depth': 5}. Best is trial 15 with value: 0.9845842587594351.


Trial 12: Accuracy = 0.9123792080675137
Trial 4: Accuracy = 0.9123792080675137


[I 2025-03-17 05:04:23,985] Trial 2 finished with value: 0.8878865381822691 and parameters: {'n_estimators': 100, 'learning_rate': 0.01, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 2: Accuracy = 0.8878865381822691


[I 2025-03-17 05:04:46,795] Trial 9 finished with value: 0.9260041213242574 and parameters: {'n_estimators': 100, 'learning_rate': 0.1, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 9: Accuracy = 0.9260041213242574


[I 2025-03-17 05:31:53,505] Trial 18 finished with value: 0.8878865381822691 and parameters: {'n_estimators': 100, 'learning_rate': 0.01, 'max_depth': 7}. Best is trial 15 with value: 0.9845842587594351.


Trial 18: Accuracy = 0.8878865381822691
Best Parameters: {'n_estimators': 50, 'learning_rate': 0.5, 'max_depth': 3}


El mejor trail ha sido el trail 15. Que nos ha dado una precision de 98,46%. Hemos conseguido esto con los siguiente parametros:
- n_estimators = 50
- learning_rate = 0.5
- max_depth = 3

Voy a probar esto

#### Probamos los mejores resultado

In [ ]:
# Definir el modelo de clasificacion de GradientBoosting
model = GradientBoostingClassifier(n_estimators=50, learning_rate=0.5,max_depth=3,random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train_norm,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.06850078955185766
Accuracy: 0.9399055845406348


Precision de 93,99%. No es el resultado esperado pero si el mas alto hasta ahora

#### Viendo esto, buscamos variacion breve en los parametros para ver si podemos sacar alguna decima mas

In [24]:
# Definir el modelo de clasificacion de GradientBoosting
model = GradientBoostingClassifier(n_estimators=45, learning_rate=0.5,max_depth=3,random_state=42)

print("Training: ")
# Entrenarlo
model.fit(X_awid_train_norm,y_awid_train)

# Obtener las predicciones para el conjunto de datos de testeo
y_est = model.predict(X_awid_test_norm)

# Obtener el error cuadratico medio y el precision
MSE_tst = mean_squared_error(y_awid_test,y_est)
accuracy = balanced_accuracy_score(y_awid_test,y_est)

# Imprimimos el error y precision del modelo
print('MSE : ' + str(MSE_tst))
print('Accuracy: ' + str(accuracy))

Training: 
MSE : 0.00989849611651666
Accuracy: 0.9445906753098985


Precision de 94,45%. Parece que si disminuyo el numero de estimadores hasta cierto punto, conseguire aumentara la precision asi que pruebo con algo mas

# Resultado Final

La mejor precision ha sido con optuna que esta fuera de scikit-learn que nos ha dado un 98,46%. Sin embargo al no se hasta que punto este resultado es fiable asi por ello pasaremos a la siguiente precision mas alta que es 94,45%